File: mo-system/mo-system-service/src/main/java/ink/ckx/mo/system/service/impl/SysRoleServiceImpl.java
package ink.ckx.mo.system.service.impl;

import cn.hutool.core.collection.CollUtil;
import cn.hutool.core.lang.Assert;
import cn.hutool.core.util.StrUtil;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.LambdaUpdateWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import ink.ckx.mo.common.core.constant.CoreConstant;
import ink.ckx.mo.common.core.exception.BusinessException;
import ink.ckx.mo.common.core.result.ResultCode;
import ink.ckx.mo.common.security.util.SecurityUtil;
import ink.ckx.mo.common.web.model.Option;
import ink.ckx.mo.system.api.model.entity.SysRole;
import ink.ckx.mo.system.api.model.entity.SysRoleMenu;
import ink.ckx.mo.system.api.model.entity.SysUserRole;
import ink.ckx.mo.system.api.model.form.RoleForm;
import ink.ckx.mo.system.api.model.query.RolePageQuery;
import ink.ckx.mo.system.api.model.vo.role.RoleDetailVO;
import ink.ckx.mo.system.api.model.vo.role.RolePageVO;
import ink.ckx.mo.system.converter.RoleConverter;
import ink.ckx.mo.system.mapper.SysRoleMapper;
import ink.ckx.mo.system.service.SysMenuService;
import ink.ckx.mo.system.service.SysRoleMenuService;
import ink.ckx.mo.system.service.SysRoleService;
import ink.ckx.mo.system.service.SysUserRoleService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * 角色业务实现类
 *
 * @author chenkaixin
 */
@Service
@RequiredArgsConstructor
public class SysRoleServiceImpl extends ServiceImpl<SysRoleMapper, SysRole> implements SysRoleService {

    private final SysRoleMenuService sysRoleMenuService;

    private final SysUserRoleService sysUserRoleService;

    private final RoleConverter roleConverter;

    private final SysMenuService menuService;

    /**
     * 角色分页列表
     *
     * @param rolePageQuery
     * @return
     */
    @Override
    public Page<RolePageVO> listRolePages(RolePageQuery rolePageQuery) {
        // 查询参数
        int pageNum = rolePageQuery.getPageNum();
        int pageSize = rolePageQuery.getPageSize();
        String keywords = rolePageQuery.getKeywords();

        // 查询数据
        Page<SysRole> rolePage = this.page(new Page<>(pageNum, pageSize),
                new LambdaQueryWrapper<SysRole>()
                        .and(StrUtil.isNotBlank(keywords), wrapper -> wrapper.like(StrUtil.isNotBlank(keywords),
                                SysRole::getName, keywords).or().like(StrUtil.isNotBlank(keywords),
                                SysRole::getCode, keywords))
                        // 非超级管理员不显示超级管理员角色
                        .ne(!SecurityUtil.isSuperAdmin(), SysRole::getCode, CoreConstant.SUPER_ADMIN_CODE)
        );

        // 实体转换
        return roleConverter.entity2Page(rolePage);
    }

    /**
     * 角色下拉列表
     *
     * @return
     */
    @Override
    public List<Option<Long>> listRoleOptions() {
        // 查询数据
        List<SysRole> roleList = this.list(new LambdaQueryWrapper<SysRole>()
                .select(SysRole::getId, SysRole::getName)
                .ne(!SecurityUtil.isSuperAdmin(), SysRole::getCode, CoreConstant.SUPER_ADMIN_CODE)
                .orderByAsc(SysRole::getSort));

        // 实体转换
        return roleConverter.roles2Options(roleList);
    }

    @Override
    public RoleDetailVO getRoleDetail(Long roleId) {
        SysRole entity = this.getOne(new LambdaQueryWrapper<SysRole>()
                .select(SysRole::getId, SysRole::getDataScope, SysRole::getName, SysRole::getCode, SysRole::getSort, SysRole::getStatus)
                .eq(SysRole::getId, roleId));

        return roleConverter.entity2DetailVO(entity);
    }

    /**
     * @param roleForm
     * @return
     */
    @Override
    public Long saveRole(RoleForm roleForm) {
        String roleCode = roleForm.getCode();

        long count = this.count(new LambdaQueryWrapper<SysRole>()
                .and(wrapper -> wrapper.eq(SysRole::getCode, roleCode)
                        .or()
                        .eq(SysRole::getName, roleCode)));
        Assert.isFalse(count > 0, () -> new BusinessException(ResultCode.PARAM_ERROR, "角色名称或角色编码重复，请检查！"));

        // 实体转换
        SysRole sysRole = roleConverter.form2Entity(roleForm);

        boolean result = this.save(sysRole);
        if (result) {
            menuService.refreshRolePerm();
        }
        return sysRole.getId();
    }

    @Override
    public void updateRole(Long id, RoleForm roleForm) {
        String roleCode = roleForm.getCode();

        long count = this.count(new LambdaQueryWrapper<SysRole>()
                .ne(SysRole::getId, id)
                .and(wrapper -> wrapper.eq(SysRole::getCode, roleCode)
                        .or()
                        .eq(SysRole::getName, roleCode)));
        Assert.isFalse(count > 0, () -> new BusinessException(ResultCode.PARAM_ERROR, "角色名称或角色编码重复，请检查！"));

        // 实体转换
        SysRole sysRole = roleConverter.form2Entity(roleForm);
        sysRole.setId(id);
        boolean result = this.updateById(sysRole);
        if (result) {
            menuService.refreshRolePerm();
        }
    }

    /**
     * 修改角色状态
     *
     * @param roleId
     * @param status
     * @return
     */
    @Override
    public void updateRoleStatus(Long roleId, Integer status) {
        boolean result = this.update(new LambdaUpdateWrapper<SysRole>()
                .eq(SysRole::getId, roleId).set(SysRole::getStatus, status));
        if (result) {
            menuService.refreshRolePerm();
        }
    }

    /**
     * 批量删除角色
     *
     * @param ids
     * @return
     */
    @Transactional(rollbackFor = Exception.class)
    @Override
    public void deleteRoles(String ids) {
        List<Long> roleIdList = StrUtil.split(ids, ',', 0, true, Long::valueOf);
        if (CollUtil.isEmpty(roleIdList)) {
            return;
        }
        roleIdList.forEach(id -> {
            long count = sysUserRoleService.count(new LambdaQueryWrapper<SysUserRole>().eq(SysUserRole::getRoleId, id));
            Assert.isFalse(count > 0, () -> new BusinessException(ResultCode.PARAM_ERROR, "该角色已分配用户，无法删除"));
            sysRoleMenuService.remove(new LambdaQueryWrapper<SysRoleMenu>().eq(SysRoleMenu::getRoleId, id));
        });

        boolean result = this.removeByIds(roleIdList);
        if (result) {
            menuService.refreshRolePerm();
        }
    }

    /**
     * 获取角色的资源ID集合,资源包括菜单和权限
     *
     * @param roleId
     * @return
     */
    @Override
    public List<Long> getRoleMenuIds(Long roleId) {
        // 获取角色拥有的菜单ID集合
        return sysRoleMenuService.listMenuIdsByRoleId(roleId);
    }

    /**
     * 修改角色的资源权限
     *
     * @param roleId
     * @param menuIds
     * @return
     */
    @Override
    @Transactional(rollbackFor = Exception.class)
    public void updateRoleMenus(Long roleId, List<Long> menuIds) {
        // 删除角色菜单
        sysRoleMenuService.remove(new LambdaQueryWrapper<SysRoleMenu>().eq(SysRoleMenu::getRoleId, roleId));
        // 新增角色菜单
        if (CollUtil.isNotEmpty(menuIds)) {
            List<SysRoleMenu> roleMenus = menuIds.stream()
                    .map(menuId -> new SysRoleMenu(null, roleId, menuId)).collect(Collectors.toList());
            sysRoleMenuService.saveBatch(roleMenus);
        }
        menuService.refreshRolePerm();
    }
}

File: mo-admin/mo-admin-service/src/main/kotlin/ink/ckx/mo/admin/service/impl/SysRoleServiceImpl.kt
package ink.ckx.mo.admin.service.impl

import cn.hutool.core.lang.Assert
import com.baomidou.mybatisplus.extension.kotlin.KtQueryChainWrapper
import com.baomidou.mybatisplus.extension.kotlin.KtUpdateChainWrapper
import com.baomidou.mybatisplus.extension.plugins.pagination.Page
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import ink.ckx.mo.admin.api.model.entity.SysRole
import ink.ckx.mo.admin.api.model.entity.SysRoleMenu
import ink.ckx.mo.admin.api.model.entity.SysUserRole
import ink.ckx.mo.admin.api.model.form.RoleForm
import ink.ckx.mo.admin.api.model.query.RolePageQuery
import ink.ckx.mo.admin.api.model.vo.role.RoleDetailVO
import ink.ckx.mo.admin.api.model.vo.role.RolePageVO
import ink.ckx.mo.admin.converter.RoleConverter
import ink.ckx.mo.admin.mapper.SysRoleMapper
import ink.ckx.mo.admin.service.SysMenuService
import ink.ckx.mo.admin.service.SysRoleMenuService
import ink.ckx.mo.admin.service.SysRoleService
import ink.ckx.mo.common.core.constant.CoreConstant
import ink.ckx.mo.common.core.exception.BusinessException
import ink.ckx.mo.common.core.result.ResultCode
import ink.ckx.mo.common.security.util.isSuperAdmin
import ink.ckx.mo.common.web.enums.StatusEnum
import ink.ckx.mo.common.web.model.Option
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


/**
 * 角色业务实现类
 *
 * @author chenkaixin
 */
@Service
class SysRoleServiceImpl(
    private val sysRoleMenuService: SysRoleMenuService,
    private val roleConverter: RoleConverter,
    private val menuService: SysMenuService,
) : ServiceImpl<SysRoleMapper, SysRole>(), SysRoleService {


    /**
     * 角色分页列表
     *
     * @param rolePageQuery
     * @return
     */
    override fun listRolePages(rolePageQuery: RolePageQuery): Page<RolePageVO> {
        // 查询参数
        val pageNum = rolePageQuery.pageNum
        val pageSize = rolePageQuery.pageSize
        val keywords = rolePageQuery.keywords

        // 查询数据
        val rolePage = ktQuery()
            .like(!keywords.isNullOrBlank(), SysRole::name, keywords)
            // 非超级管理员不显示超级管理员角色
            .ne(!isSuperAdmin(), SysRole::code, CoreConstant.SUPER_ADMIN_CODE)
            .orderByDesc(SysRole::sort, SysRole::updateTime)
            .page(Page(pageNum, pageSize))

        // 实体转换
        return roleConverter.entity2Page(rolePage)
    }

    /**
     * 角色下拉列表
     *
     * @return
     */
    override fun listRoleOptions(): List<Option<Long>> {
        // 查询数据
        val roleList = ktQuery()
            .select(SysRole::id, SysRole::name)
            .ne(!isSuperAdmin(), SysRole::code, CoreConstant.SUPER_ADMIN_CODE)
            .orderByAsc(SysRole::sort)
            .list()

        // 实体转换
        return roleConverter.roles2Options(roleList)
    }

    override fun getRoleDetail(roleId: Long): RoleDetailVO {
        val entity = ktQuery()
            .select(
                SysRole::id,
                SysRole::dataScope,
                SysRole::name,
                SysRole::code,
                SysRole::sort,
                SysRole::status
            )
            .eq(SysRole::id, roleId)
            .one()

        return roleConverter.entity2DetailVO(entity)
    }

    /**
     * @param roleForm
     * @return
     */
    override fun saveRole(roleForm: RoleForm): Long? {
        val roleCode = roleForm.code
        val count = ktQuery()
            .eq(SysRole::code, roleCode)
            .count()

        Assert.isFalse(count > 0) { BusinessException(ResultCode.PARAM_ERROR, "角色编码重复，请检查！") }

        // 实体转换
        val sysRole = roleConverter.form2Entity(roleForm)
        val result = save(sysRole)
        if (result) {
            menuService.refreshRolePerm()
        }
        return sysRole.id
    }

    override fun updateRole(roleId: Long, roleForm: RoleForm) {
        val roleCode = roleForm.code
        val count = ktQuery()
            .ne(SysRole::id, roleId)
            .eq(SysRole::code, roleCode)
            .count()

        Assert.isFalse(count > 0) { BusinessException(ResultCode.PARAM_ERROR, "角色编码重复，请检查！") }

        // 实体转换
        val sysRole = roleConverter.form2Entity(roleForm)
        sysRole.id = roleId
        val result = updateById(sysRole)
        if (result) {
            menuService.refreshRolePerm()
        }
    }

    /**
     * 修改角色状态
     *
     * @param roleId
     * @param status
     * @return
     */
    override fun updateRoleStatus(roleId: Long, status: StatusEnum) {
        val result = ktUpdate()
            .eq(SysRole::id, roleId)
            .set(SysRole::status, status)
            .update()
        if (result) {
            menuService.refreshRolePerm()
        }
    }

    /**
     * 批量删除角色
     *
     * @param ids
     * @return
     */
    @Transactional(rollbackFor = [Exception::class])
    override fun deleteRoles(ids: String) {
        val roleIdList = ids.split(',').mapNotNull { it.toLongOrNull() }
        if (roleIdList.isEmpty()) {
            return
        }
        roleIdList.forEach {
            val count = KtQueryChainWrapper(SysUserRole()).eq(SysUserRole::roleId, it).count()
            Assert.isFalse(count > 0) { BusinessException(ResultCode.PARAM_ERROR, "该角色已分配用户，无法删除") }
            KtUpdateChainWrapper(SysRoleMenu()).eq(SysRoleMenu::roleId, it).remove()
        }
        val result = removeByIds(roleIdList)
        if (result) {
            menuService.refreshRolePerm()
        }
    }

    /**
     * 获取角色的资源ID集合,资源包括菜单和权限
     *
     * @param roleId
     * @return
     */
    override fun getRoleMenuIds(roleId: Long): List<Long> {
        // 获取角色拥有的菜单ID集合
        return sysRoleMenuService.listMenuIdsByRoleId(roleId)
    }

    /**
     * 修改角色的资源权限
     *
     * @param roleId
     * @param menuIds
     * @return
     */
    @Transactional(rollbackFor = [Exception::class])
    override fun updateRoleMenus(roleId: Long, menuIds: List<Long>) {
        // 删除角色菜单
        KtUpdateChainWrapper(SysRoleMenu()).eq(SysRoleMenu::id, roleId).remove()
        // 新增角色菜单
        if (menuIds.isNotEmpty()) {
            val roleMenus = menuIds.map { SysRoleMenu(null, roleId, it) }.toList()
            sysRoleMenuService.saveBatch(roleMenus)
        }
        menuService.refreshRolePerm()
    }
}